File: examples/todomvc/src/todos.js
import Type from 'union-type';
import { bind, sequence, isBoolean, classnames } from './helpers';
import task from './task';

const KEY_ENTER = 13;

// model : { nextID: Number, editingTitle: String, tasks: [task.model], filter: String }
const Action = Type({
  Add: [String],
  Remove: [Number],
  Archive: [],
  ToggleAll: [isBoolean],
  Filter: [String],
  Modify: [Number, task.Action],
});

const targetChecked = e => Boolean(e.target.checked);
// const targetValue = e => e.target.value;

function onInput(handler, e) {
  if (e.keyCode === KEY_ENTER && e.target.value !== '') {
    handler(Action.Add(e.target.value));
  }
}

function remainingTodos(tasks) {
  // eslint-disable-next-line
  return tasks.reduce((acc, task) => !task.done ? acc + 1 : acc, 0);
}

function filteredTodos(tasks, filter) {
  /* eslint-disable */
  return filter === 'completed' ? tasks.filter(todo => todo.done)
         : filter === 'active'  ? tasks.filter(todo => !todo.done)
                                : tasks;
  /* eslint-enable */
}

function view(h, model, handler) {
  const remaining = remainingTodos(model.tasks);
  const filtered = filteredTodos(model.tasks, model.filter);

  window.onhashchange = () => handler(Action.Filter(window.location.hash.substr(2) || 'all'));

  return h('section', {
    className: 'todoapp',
    raw: {
      onclick: () => console.log('clicked'),
    },
  }, [
    h('header', {
      className: 'header',
    }, [
      h('h1', 'todos'),
      h('input', {
        id: 'new-todo',
        className: 'new-todo',
        placeholder: 'What needs to be done?',
        raw: {
          value: model.editingTitle,
          onkeydown: bind(onInput, handler),
        },
      }),
    ]),
    h('section', {
      className: 'main',
      style: `display: ${model.tasks.length ? 'block' : 'none'}`,
    }, [
      h('input', {
        className: 'toggle-all',
        type: 'checkbox',
        raw: {
          checked: remaining === 0,
          onclick: sequence(targetChecked, Action.ToggleAll, handler),
        },
      }),
      h('ul', {
        className: 'todo-list',
      }, filtered.map(todo =>
        task.view(
          h,
          todo,
          action => handler(Action.Modify(todo.id, action)),
          id => handler(Action.Remove(id)),
        ))),
    ]),
    h('footer', {
      className: 'footer',
      style: `display: ${model.tasks.length ? 'block' : 'none'}`,
    }, [
      h('span', {
        className: 'todo-count',
      }, [
        h('strong', String(remaining)),
        ` item${remaining === 1 ? '' : 's'} left`,
      ]),
      h('ul', {
        className: 'filters',
      }, [
        h('li', [h('a', { className: classnames({ selected: model.filter === 'all' }), href: '#/' }, 'All')]),
        h('li', [h('a', { className: classnames({ selected: model.filter === 'active' }), href: '#/active' }, 'Active')]),
        h('li', [h('a', { className: classnames({ selected: model.filter === 'completed' }), href: '#/completed' }, 'Completed')]),
      ]),
      h('button', {
        className: 'clear-completed',
        style: `display: ${model.tasks.length !== remaining ? 'block' : 'none'}`,
        raw: { onclick: bind(handler, Action.Archive()) },
      }, 'Clear completed'),
    ]),
  ]);
}

function init(tasks = []) {
  return {
    // eslint-disable-next-line
    nextID: tasks.reduce((acc, task) => Math.max(acc, task.id), 0) + 1,
    tasks,
    editingTitle: '',
    filter: 'all',
  };
}

function addTodo(model, title) {
  return {
    ...model,
    tasks: [
      ...model.tasks,
      task.init(model.nextID, title),
    ],
    editingTitle: '',
    nextID: model.nextID + 1,
  };
}

function removeTodo(model, id) {
  return {
    ...model,
    // eslint-disable-next-line
    tasks : model.tasks.filter(taskModel => taskModel.id !== id)
  };
}

function archiveTodos(model) {
  return {
    ...model,
    // eslint-disable-next-line
    tasks : model.tasks.filter( taskModel => !taskModel.done),
  };
}

function toggleAll(model, done) {
  return {
    ...model,
    // eslint-disable-next-line
    tasks : model.tasks.map(taskModel => task.update(taskModel, task.Action.Toggle(done)))
  };
}


function modifyTodo(model, id, action) {
  return {
    ...model,
    tasks: model.tasks.map(
      // eslint-disable-next-line
      taskModel => taskModel.id !== id ? taskModel : task.update(taskModel, action)
    ),
  };
}

function update(model, action) {
  return Action.case({
    Add: title => addTodo(model, title),
    Remove: id => removeTodo(model, id),
    Archive: () => archiveTodos(model),
    ToggleAll: done => toggleAll(model, done),
    Filter: filter => ({ ...model, filter }),
    Modify: (id, action2) => modifyTodo(model, id, action2),
  }, action);
}

export default { view, init, update, Action };

File: examples/todomvc - cpp/src/todos.cpp
#include "./todos.hpp"
#include "./task.hpp"
#include "./helpers.hpp"
#include <emscripten/val.h>
#include <functional>
#include <vector>
#include <string>

using namespace todomvc::helpers;

namespace todomvc {

	namespace todos {

		namespace Action {

			action Add(std::string title) {
				return [title](Todos model) -> Todos {
					model.tasks.push_back(todomvc::task::init(model.nextID, title));
					model.editingTitle = std::string("");
					++model.nextID;
					return model;
				};
			};

			action Remove(int id) {
				return [id](Todos model) -> Todos {
					model.tasks.erase(
						std::remove_if(
							model.tasks.begin(),
							model.tasks.end(),
							[&id](const Task& o) { return o.id == id; }
						),
						model.tasks.end()
					);
					return model;
				};
			};

			action Archive() {
				return [](Todos model) -> Todos {
					model.tasks.erase(
						std::remove_if(
							model.tasks.begin(),
							model.tasks.end(),
							[](const Task& o) { return o.done; }
						),
						model.tasks.end()
					);
					return model;
				};
			};

			action ToggleAll(bool done) {
				return [done](Todos model) -> Todos {
					for(std::vector<Task>::size_type i = 0; i < model.tasks.size(); ++i) {
						model.tasks[i] = todomvc::task::update(model.tasks[i], todomvc::task::Action::Toggle(done));
					}
					return model;
				};
			};

			action Filter(TodoFilter filter) {
				return [filter](Todos model) -> Todos {
					model.filter = filter;
					return model;
				};
			};

			action Modify(int id, todomvc::task::action act) {
				return [act, id](Todos model) -> Todos {
					for(std::vector<Task>::size_type i = 0; i < model.tasks.size(); ++i) {
						if (model.tasks[i].id == id) {
							model.tasks[i] = todomvc::task::update(model.tasks[i], act);
						}
					}
					return model;
				};
			};

		}

		const int KEY_ENTER = 13;

		bool onClick(emscripten::val e) {
			emscripten::val::global("console").call<void>("log", emscripten::val("clicked"));
			return true;
		};

		int remainingTodos(const std::vector<Task>& tasks) {
			int remaining = 0;
			for(std::vector<Task>::size_type i = 0; i < tasks.size(); ++i) {
				if (!tasks[i].done) {
					++remaining;
				}
			}
			return remaining;
		};

		std::vector<Task> filteredTodos(const std::vector<Task>& tasks, const TodoFilter& filter) {
			if (filter == all) return tasks;

			std::vector<Task> filtered = std::vector<Task>();
			for(std::vector<Task>::size_type i = 0; i < tasks.size(); ++i) {
				switch(filter) {
						case completed:
							if (tasks[i].done) filtered.push_back(tasks[i]);
							break;
						case active:
							if (!tasks[i].done) filtered.push_back(tasks[i]);
							break;
						default:
							break;
				}
			}
			return filtered;
		};

		VNode* view(Todos model, std::function<void(action)> handler) {
			int remaining = remainingTodos(model.tasks);
			std::vector<Task> filtered = filteredTodos(model.tasks, model.filter);

			Children tasksViews = Children();
			for(std::vector<Task>::size_type i = 0; i < filtered.size(); ++i) {
				Task todo = filtered[i];
				tasksViews.push_back(
					todomvc::task::view(
						todo,
						[handler, todo](std::function<Task(Task)> action) -> void {
							return handler(Action::Modify(todo.id, action));
						},
						[handler](int id) -> void {
							return handler(Action::Remove(id));
						}
					)
				);
			}

			return h("section",
				Data(
					Attrs {
						{"class", "todoapp"}
					},
					Callbacks {
						{"onclick", onClick}
					}
				),
				Children {
					h("header",
						Data(
							Attrs {
								{"class", "header"}
							}
						),
						Children {
							h(std::string("h1"), std::string("todos")),
							h("input",
								Data(
									Attrs {
										{"id", "new-todo"},
										{"class", "new-todo"},
										{"placeholder", "What needs to be done?"}
									},
									Props {
										{"value", emscripten::val(model.editingTitle)}
									},
									Callbacks {
										{"onkeydown", [handler](emscripten::val e) -> bool {
											std::string value = wstring_to_utf8(e["target"]["value"].as<std::wstring>());
											if (e["keyCode"].as<int>() == KEY_ENTER && !value.empty()) {
												handler(Action::Add(value));
											}
											return true;
										}}
									}
								)
							)
						}
					),
					h("section",
						Data(
							Attrs {
								{"class", "main"},
								{"style", "display: " + std::string(model.tasks.size() > 0 ? "block" : "none")}
							}
						),
						Children {
							h("input",
								Data(
									Attrs {
										{"class", "toggle-all"},
										{"type", "checkbox"}
									},
									Props {
										{"checked", emscripten::val(remaining == 0)}
									},
									Callbacks {
										{"onclick", [handler](emscripten::val e) -> bool {
											handler(Action::ToggleAll(targetChecked(e)));
											return true;
										}}
									}
								)
							),
							h("ul",
								Data(
									Attrs {
										{"class", "todo-list"}
									}
								),
								tasksViews
							)
						}
					),
					h("footer",
						Data(
							Attrs {
								{"class", "footer"},
								{"style", "display: " + std::string(model.tasks.size() > 0 ? "block" : "none")}
							}
						),
						Children {
							h("span",
								Data(
									Attrs {
										{"class", "todo-count"}
									}
								),
								Children {
									h(std::string("strong"), std::to_string(remaining)),
									h(std::string(" item" + std::string(remaining == 1 ? "" : "s") + " left"), true)
								}
							),
							h("ul",
								Data(
									Attrs {
										{"class", "filters"}
									}
								),
								Children {
									h("li",
										h("a",
											Data(
												Attrs {
													{"class", model.filter == all ? "selected" : ""},
													{"href", "#/"}
												}
											),
											std::string("All")
										)
									),
									h("li",
										h("a",
											Data(
												Attrs {
													{"class", model.filter == active ? "selected" : ""},
													{"href", "#/active"}
												}
											),
											std::string("Active")
										)
									),
									h("li",
										h("a",
											Data(
												Attrs {
													{"class", model.filter == completed ? "selected" : ""},
													{"href", "#/completed"}
												}
											),
											std::string("Completed")
										)
									)
								}
							),
							h("button",
								Data(
									Attrs {
										{"class", "clear-completed"},
										{"style", "display: " + std::string(model.tasks.size() > 0 ? "block" : "none")}
									},
									Callbacks {
										{"onclick", [handler](emscripten::val e) -> bool {
											handler(Action::Archive());
											return true;
										}}
									}
								),
								std::string("Clear completed")
							)
						}
					)
				}
			);
		};

		Todos init() {
			return init(std::vector<Task>());
		};

		Todos init(std::vector<Task> tasks) {
			int nextID = 0;
			for (auto it = tasks.begin(); it != tasks.end(); ++it) {
				if (nextID < it->id) {
					nextID = it->id;
				}
			}
			++nextID;
			return {
				.nextID = nextID,
				.tasks = tasks,
				.editingTitle = std::string(""),
				.filter = all,
			};
		};

		Todos update(Todos todos, action act) {
			return act(todos);
		};

	}

}
